### PR TITLE
Start out the enchannel document

### PR DESCRIPTION
**Pull request by rgbkrk**
This gets us started in providing developer guides for working with Jupyter via enchannel, explaining the motivations behind kernels and a little about how we communicate with them. This is one of several outlined in https://github.com/nteract/nteract/issues/182.

Once this is a little more ready, I'll rip the WIP tag off.

Note: I'm focused on getting some kind of content up before doing site layout.

### CODE REVIEW

**Comment by rgbkrk** [comment]
Alright, this is a little bit of a start. It definitely needs more work to help with the flow, defining more, and including diagrams.

**Comment by ivanov** [comment]
WIP tag not there, so merging! :+1:

**Comment by betatim** [comment]
Nice tutorial. I followed along and gained knowledge :book:

File: enchannel/build-your-own-repl.md [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+## Requirements
+
+First thing you'll need for this tutorial is the IPython kernel for Python3.
+
+```
+python3 -m pip install ipykernel
+python3 -m ipykernel install --user
+```
+
+
```
Comment (jdfreder) [inline enchannel/build-your-own-repl.md]: `You'll also need:`

File: enchannel/build-your-own-repl.md [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+* [Node.js 5+ and npm 3](https://nodejs.org/en/)
+* ZeroMQ
+* Python 2 (for builds - you can still run Python 3)
+* IPython kernels installed
+* `git`
+
+Each operating system has their own instruction set to get node zmq working. Please read on down to save yourself time.
```
Comment (jdfreder) [inline enchannel/build-your-own-repl.md]: s/has their/has it's

File: enchannel/build-your-own-repl.md [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+
+Note: If you're gripping tightly to an older node version, try out
+[nvm](https://github.com/creationix/nvm) or [n](https://github.com/tj/n) for all
+your node environment switching needs. As for zmq,
+Once [zmq-prebuilt](https://github.com/nteract/zmq-prebuilt) is ready for
```
Comment (jdfreder) [inline enchannel/build-your-own-repl.md]: Whoops no need for capitalization here

File: enchannel/build-your-own-repl.md [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+This guide is here to help you on your journey to working with nteract's node
+packages. The main goal of this guide is to go from scratch to a working interactive
+[REPL](https://en.wikipedia.org/wiki/Read%E2%80%93eval%E2%80%93print_loop) you can
+use at the command line.
+
+Core topics that will be introduced include:
+
+* Jupyter kernels
+* Messaging with Jupyter (via ZeroMQ)
+* Observables
```
Comment (willingc) [inline enchannel/build-your-own-repl.md]: OK I'll bite. What's an observable?

File: enchannel/build-your-own-repl.md [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+
+```
+python3 -m pip install ipykernel
+python3 -m ipykernel install --user
+```
+
+
+* [Node.js 5+ and npm 3](https://nodejs.org/en/)
+* ZeroMQ
+* Python 2 (for builds - you can still run Python 3)
```
Comment (willingc) [inline enchannel/build-your-own-repl.md]: for builds Python 2 is needed; you may still run either Python 2 or Python 3

File: enchannel/build-your-own-repl.md [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+* IPython kernels installed
+* `git`
+
+Each operating system has their own instruction set to get node zmq working. Please read on down to save yourself time.
+
+Note: If you're gripping tightly to an older node version, try out
+[nvm](https://github.com/creationix/nvm) or [n](https://github.com/tj/n) for all
+your node environment switching needs. As for zmq,
+Once [zmq-prebuilt](https://github.com/nteract/zmq-prebuilt) is ready for
+everyone, you won't have to do these steps
```
Comment (willingc) [inline enchannel/build-your-own-repl.md]: Punctuation at end of sentence needed

File: enchannel/build-your-own-repl.md [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+
+### OS X
+
+#### homebrew on OS X
```
Comment (willingc) [inline enchannel/build-your-own-repl.md]: Homebrew

File: enchannel/build-your-own-repl.md [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+
+```
+npm init
+```
+
+Feel free to accept the defaults or mix it up. Go ahead and add your new `package.json` to `git`:
+
+```
+git add package.json
+```
```
Comment (willingc) [inline enchannel/build-your-own-repl.md]: Perhaps obvious to most, but likely helpful to point out that git commit is needed too.

File: enchannel/build-your-own-repl.md [hunk 9]
```diff
@@ -0,0 +1,411 @@
...
+In [2]: random.random()
+Out[2]: 0.037744852607581425
+
+In [3]:
+```
+
+It functions like a REPL by allowing a user to provide input and get responses back. What's different is that the kernels also provide a rich cross-language protocol
+for sending back data as text, HTML, images, etc. while also providing a means to do things like tab completion.
+
+A kernel always operates in a separate process from the application that launches it. Here, `spawnteract` does the launching for us, spawning the process (`kernel.spawn`) as well as creating a connection file (`kernel.connectionFile` is the file, `kernel.config` are the contents) to allow frontends to connect to it.
```
Comment (willingc) [inline enchannel/build-your-own-repl.md]: ...creating a connection file, `kernel.connectionFile`, which has `kernel.config` as its content to allow ...

File: enchannel/build-your-own-repl.md [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+  },
+  metadata: {},
+  parent_header: {},
+  content: {},
+};
+```
+
+This is probably lower level than you would expect to work with at the get-go and we'll do our best to create helper functions along the way. Part of the adventure here is learning the Jupyter protocol and formats. :smile:
+
+The next item we want to setup is a listener for what will be the response to our message:
```
Comment (willingc) [inline enchannel/build-your-own-repl.md]: ...a listener for our message response:

**Comment by willingc** [comment]
This is great!

File: enchannel/build-your-own-repl.md (hunk 4 above)
Comment (rgbkrk) [inline enchannel/build-your-own-repl.md]: The best description I have is that it's an asynchronously flowing data stream. Should I provide a link at this point in the doc?

**Comment by rgbkrk** [comment]
Oh my! Look at this feedback. Thank you!

**Comment by rgbkrk** [comment]
I'm wrapping up some nteract/nteract work tonight. If no one takes on these edits, I'll tackle doc updates tomorrow night.

File: enchannel/build-your-own-repl.md (hunk 2 above)
Comment (minrk) [inline enchannel/build-your-own-repl.md]: s/it's/its/

File: enchannel/build-your-own-repl.md (hunk 4 above)
Comment (willingc) [inline enchannel/build-your-own-repl.md]: If there is a place to link, then sure. Otherwise, it's fine for now.

**Pull request merged**